### PR TITLE
Automatic identification of Boundaries based on continuity

### PR DESCRIPTION
**Pull request by jzwar**
# Overview
The idea is simple, if two adjacent elements share a boundary that connect G1-continouous, they probably belong to the same boundary on a global scale. This hopefully facilitates the application of boundary conditions in IGA.

## Addressed issues
- #81 

## Showcase
TBD

## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [x] Added test(s)

### CODE REVIEW

**Review by j042 — COMMENTED** [review]
Good stuff

File: cpp/splinepy/py/py_spline_exporter.hpp
Comment (jzwar) [reply cpp/splinepy/py/py_spline_exporter.hpp:1245]: Done

File: splinepy/multipatch.py
Comment (jzwar) [reply splinepy/multipatch.py]: Done

File: splinepy/multipatch.py
Comment (jzwar) [reply splinepy/multipatch.py]: Done

File: splinepy/multipatch.py
Comment (jzwar) [reply splinepy/multipatch.py:203]: I wanted to keep the interfaces as easy as possible. So making this a property was intentional. In the interface property it gets computed automatically if it is not yet provided. Here, it makes no sense to provide the interface boundaries, as they are a function of the interfaces. 

Can we maybe push that to the next pr?

File: splinepy/multipatch.py
Comment (jzwar) [reply splinepy/multipatch.py:228]: I think there is no need to adapt the tolerances here, as the boundary interfaces are only really required for the `boundary_from_continuity` function. All the cost intensive property calls are just needed there. So I think it is more important to keep things simple here

**Comment by jzwar** [comment]
I just realized, that my approach was kind of stupid. As boundary patches I could return a Multipatch object and reuse all the code :facepalm:

**Comment by jzwar** [comment]
image

It seems to work properly, although the plotting is still wip

File: cpp/splinepy/py/py_spline_exporter.hpp
Comment (j042) [reply cpp/splinepy/py/py_spline_exporter.hpp:1186]: alright, I will implement some list-like container for multi-patch splines

**Review by j042 — CHANGES_REQUESTED** [review]

File: splinepy/multipatch.py
Comment (j042) [reply splinepy/multipatch.py]: I don't think defining a function with `@property` make it necessarily easy. It can be an empty arg function call.

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:211]: This doesn't work well with inplace changes

File: splinepy/multipatch.py
Comment (j042) [reply splinepy/multipatch.py:202]: keep it as a normal function for the future?

File: splinepy/multipatch.py
Comment (jzwar) [reply splinepy/multipatch.py:211]: This is true, I am not yet sure how to handle the in-place changes at all, especially if splines are modified on the fly. The only safeguard I have at this point is in the splines-setter, where - if called - all dependent members are reinitialized... I have no better solution at this point

File: CMakeLists.txt
Comment (jzwar) [reply CMakeLists.txt]: Done

File: cpp/splinepy/py/py_spline_exporter.hpp
Comment (jzwar) [reply cpp/splinepy/py/py_spline_exporter.hpp]: I think I found a solution to this, check out the new exporter

File: CMakeLists.txt
Comment (j042) [inline CMakeLists.txt:33]:
```suggestion
set(SPLINEPY_DEFS $<$<NOT:$<CONFIG:Debug>>:NDEBUG>)
```

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:388]:
```suggestion
            return None
```

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:437]:
```suggestion
        return None
```

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:461]:
```suggestion
            return None
```

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:434]: make this function param?

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:335]:
```suggestion
    def boundaries_from_function(
```
or you can change all the other one to `boundary_*`

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:338]:
```suggestion
        mask=None,
```
boundaries from function from boundaries is a bit unclear

File: splinepy/multipatch.py
Comment (j042) [reply splinepy/multipatch.py:209]: I'd say, we can leave this to always recompute or add a `use_saved` param until we can handle in-place changes

**Review by j042 — COMMENTED** [review]
Some of the docstrings are incomplete (splines, interfaces, ....)
Other than that, thanks for updating!

File: splinepy/multipatch.py
Comment (jzwar) [reply splinepy/multipatch.py]: The others auto assign multiple boundaries, this one does one at a time. I reconsidered this change and changed it back, :sweat_smile:

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:426]:
```suggestion
            self._logd(
```

File: splinepy/multipatch.py
Comment (j042) [inline splinepy/multipatch.py:516]:
```suggestion
            self._logd(
```

**Review by j042 — APPROVED** [review]
Thanks for this new feature :) 
Please update those comments before merging

**Pull request merged**
